app/models/course/lesson_plan/item.rb

```diff
@@ -28,11 +28,19 @@ class Course::LessonPlan::Item < ApplicationRecord
   end)
 
   # @!method self.with_actable_types
-  #   Scopes the lesson plan items to those which belong to the given actable_types
+  #   Scopes the lesson plan items to those which belong to the given actable_types.
+  #   Each actable type is further scoped to return the IDs of items for display.
+  #   The course settings object is provided so the actable types can look up their settings to
+  #   decide if they should be displayed.
   #
   # @param actable_types [Array<String>] Array of strings with the actable type names.
-  scope :with_actable_types, lambda { |actable_types|
-    where(actable_type: actable_types)
+  # @param settings [SettingsOnRails::Settings] Course settings object.
+  scope :with_actable_types, lambda { |actable_types, settings|
```

Comment:
I don't think we want to be sending the settings object into the scope like that, it introduces a lot of coupling between component settings and the model. Changes to the implementation of settings would result in a need to change the code within the `for_display` scopes in each component.

A preferable scope signature would be: 

```ruby
scope :with_actable_type, lambda do |actable_hash|
    ....
end

# actable_hash has the following shape: 
{ 
  'Course::Assessment' => [...], # some kind of data representing the tab_ids, hash or array.
  'Course::Video' => nil, # Can be just an empty object or array or something, 
  ...
}
``` 

Something along this lines would be good, where the paramters going into the scope are simple pieces of data rather than complex objects.

Reply:
This moves the coupling to the items controller and the scope.

The controller will have to generate `actable_hash`, including the key value (array of disabled tab IDs in this case). Then the scope has to figure out what to do with that array of numbers (exclude, not include in this case).

If that's preferable I'll make the change.

Reply:
Lesson plan N+1 queries are currently dealt with in `Course::LessonPlan::ItemsController#render_json_response` where we load tab titles and materials in separate queries. This is not very clean because the controller knows what each component needs specifically.

Since the querying is now more complicated, would it be a good idea to do a single load for each model that `acts_as_lesson_plan_item` instead of trying to capture all lesson plan items in a single query here? 

In that case, instead of this scope, maybe we can have a method that asks each component that `acts_as_lesson_plan_item` for a list of items showable in the lesson plan. Those components can then go and ask their respective settings models which items should be displayed (the require settings are stored in the respective slices and not in `:course_lesson_plan_component` anyway).

What do you guys think?

Reply:
> This moves the coupling to the items controller and the scope.

Somewhat, but I think is less coupling as it releases some dependency on the settings object.

To be fair, I was hoping the changes introduced in https://github.com/Coursemology/coursemology2/pull/2694 are in the same place, but I found that some are in settings, and some are in components. If there is a way to rework part of this to make it easier that would be good - have a try first?